### PR TITLE
server: add query stats to hrana responses

### PR DESCRIPTION
**Pull request by LucioFranco**

### CODE REVIEW

**Comment by invisal** [comment]
Wow. This is game changing feature.

File: docs/HRANA_3_SPEC.md [hunk 1]
```diff
@@ -1026,6 +1026,8 @@ type StmtResult = {
     "rows": Array<Array<Value>>,
     "affected_row_count": uint32,
     "last_insert_rowid": string | null,
+    "rows_read": uint64,
```
Comment (haaawk) [inline docs/HRANA_3_SPEC.md:1029]: Since `affected_row_count` is uint32, shouldn't `rows_read` and `rows_written` be uint32 as well?

File: libsql/src/hrana/cursor.rs [hunk 2]
```diff
@@ -134,6 +134,8 @@ where
                 affected_row_count,
                 last_insert_rowid,
                 replication_index: None,
+                rows_read: 0,
+                rows_written: 0,
```
Comment (haaawk) [inline libsql/src/hrana/cursor.rs:138]: Why are those always 0 here?

File: libsql-hrana/src/proto.rs [hunk 3]
```diff
@@ -22,6 +22,9 @@ pub struct PipelineRespBody {
     pub results: Vec<StreamResult>,
 }
 
+#[derive(Serialize, Deserialize, prost::Message)]
+pub struct Stats {}
```
Comment (haaawk) [inline libsql-hrana/src/proto.rs:26]: Where is this used?

File: libsql-server/src/connection/program.rs [hunk 4]
```diff
@@ -105,7 +106,7 @@ impl<'a, B, F, S> Vm<'a, B, F, S>
 where
     B: QueryResultBuilder,
     F: Fn(&StmtKind) -> (bool, Option<String>),
-    S: Fn(String, &rusqlite::Statement, Duration),
+    S: Fn(String, u64, u64, u64, Duration),
```
Comment (haaawk) [inline libsql-server/src/connection/program.rs:109]: Why is this change needed? It seems that the whole PR would be much simpler if we kept accepting Statement here, no?

**Review by haaawk — APPROVED** [review]

File: docs/HRANA_3_SPEC.md (hunk 1 above)
Comment (LucioFranco) [reply docs/HRANA_3_SPEC.md:1029]: I am not following for why it should match the uint32 of affected row count? Is it just to save the 4 extra bytes?

File: libsql/src/hrana/cursor.rs (hunk 2 above)
Comment (LucioFranco) [reply libsql/src/hrana/cursor.rs:138]: Cursors for now don't implement stats because its too complicated and not needed for the requirements set out by the requested users.

File: libsql-hrana/src/proto.rs (hunk 3 above)
Comment (LucioFranco) [reply libsql-hrana/src/proto.rs:26]: I believe this was extra, removing it. :)

File: libsql-server/src/connection/program.rs (hunk 4 above)
Comment (LucioFranco) [reply libsql-server/src/connection/program.rs:109]: I did this change because it made the most sense, we only want to request these stats once and then pass them to the callback and to the add_stats of the build. In addition, we want these values before we hit the callback. This simplification made the most sense to me rather than leaking a detail like this.

Comment (haaawk) [reply libsql-server/src/connection/program.rs:109]: I have nothing against this change but it is completely irrelevant to the goal of this PR. It's better to do it in a separate PR as a follow up. Otherwise I needed to spend 50% of time reviewing non-essential change. No need to do anything now but in the future having independent changes in independent PRs or at least commits is a desirable thing.

File: docs/HRANA_3_SPEC.md (hunk 1 above)
Comment (haaawk) [reply docs/HRANA_3_SPEC.md:1029]: Just for consistency. if affected_row_count is uint32 and rows_written is uint64 then it is confusing. Does it mean max affected_row_count is 2^32 and max rows_written is 2^64? Maybe it's fine to have different types for those fields but it is confusing.

**Review by haaawk — APPROVED** [review]

File: docs/HRANA_3_SPEC.md (hunk 1 above)
Comment (LucioFranco) [reply docs/HRANA_3_SPEC.md:1029]: lets solve this in a follow up. This feature is still hidden for now. Looks like there is both a uint64 and uint32 api for affected rows.
